
# SDS 3386 2022 Fall Assignment 2: Data Wrangling

2022 Fall, Tanya Schmah

## Due Thursday October 6th at 6pm, on Brightspace

Total Marks: 10

This is a **group** assignment. You should complete it with two other people, preferably those you have chosen to work on a project with (though this is not required). By the time you submit your assignment, all group members should understand all answers and be able to explain them.

Start here: Download this file to your laptop, rename or make a copy of it with filename in the format: SDS3386_A2_Last1_Last2_Last3.ipynb
(In this class, last names are unique.)
Complete the assignment by editing the file. If group members work on different copies of the notebook, then when you're finished you must combine your notebooks into one "clean" copy for submission. When finished, export as HTML, and upload both the jupyter notebook and the HTML file to Brightspace. Each student should submit a copy of the same assignment to Brightspace, for easier marking.

You may copy and modify code from your Lab exercises, but if you copy more than two lines from another source, note the source in a comment.

# Question 1 (2 points)

Using the basketball data in "https://raw.githubusercontent.com/fivethirtyeight/data/master/nba-elo/nbaallelo.csv", determine the proportion of Home games ('game_location' = 'H') won, and the proportion of Away games won, over all the seasons of the NBA recorded in this dataset.

*Hint: run 'df.value_counts?'*

In [4]:
import pandas as pd  
import altair as alt
import numpy as np

In [54]:
url = "https://raw.githubusercontent.com/fivethirtyeight/data/master/nba-elo/nbaallelo.csv"
data = pd.read_csv(url)
data


gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  ...  win_equiv  opp_id  \
0                1            0     TRH    Huskies  ...  40.294830     NYK   
1                1            0     NYK     Knicks  ...  41.705170     TRH   
2                1            0     CHS      Stags  ...  42.012257     NYK   
3                2            0     NYK     Knicks  ...  40.692783     CHS   
4                1            0     DTF    Falcons  ...  38.864048     WSC   
...            ...          ...     ...        ...  ...        ...     ...   
126309         100            1     CLE  Cavaliers  ...  60.309792     GSW   
126310         102            1     GSW   Warriors  ...  68.013329     CLE   
126311         101            1     CLE  Cavaliers  ...  60.010067     GSW   
126312         102            1     CLE  Cavaliers  ...  59.290245     GSW   
126313         103            1     GSW   Warriors  ...  68.519516     CLE   

         opp_fran  opp_pts  opp_elo_i  opp_elo_n  game_location  game_result  \
0          Knicks       68  1300.0000  1306.7233              H            L   
1         Huskies       66  1300.0000  1293.2767              A            W   
2          Knicks       47  1306.7233  1297.0712              H            W   
3           Stags       63  1300.0000  1309.6521              A            L   
4        Capitols       50  1300.0000  1320.3811              H            L   
...           ...      ...        ...        ...            ...          ...   
126309   Warriors      103  1790.9591  1809.9791              H            L   
126310  Cavaliers       91  1704.3949  1700.7391              H            W   
126311   Warriors      104  1809.9791  1813.6349              A            L   
126312   Warriors      105  1813.6349  1822.2881              H            L   
126313  Cavaliers       97  1700.7391  1692.0859              A            W   

        forecast notes  
0       0.640065   NaN  
1       0.359935   NaN  
2       0.631101   NaN  
3       0.368899   NaN  
4       0.640065   NaN  
...          ...   ...  
126309  0.546572   NaN  
126310  0.765565   NaN  
126311  0.234435   NaN  
126312  0.481450   NaN  
126313  0.518550   NaN  

[126314 rows x 23 columns]

In [71]:
#Filtering the data
won = data['game_result']== 'W'
home = data['game_location']=='H' 
away = data['game_location']=='A'

won_home = data[home & won]
won_away = data[away & won]

#Computing the proportions needed
won_home_proportion = won_home.shape[0]/data.shape[0]
won_away_proportion = won_away.shape[0]/data.shape[0]




print(f"Proportion of Home games won : {won_home_proportion} ({won_home_proportion*100:.2f} %)")

print(f"Proportion of Away games won : {won_away_proportion} ({won_away_proportion*100:.2f} %)")



Proportion of Home games won : 0.31116899156071376 (31.12 %)
Proportion of Away games won : 0.1886805896416866 (18.87 %)


# Answer:

Proportion of Home games won : 0.3112 (31.12 %)

Proportion of Away games won : 0.1887 (18.87 %)




# Question 2 (4 points)

Use the dataset `grades.csv` provided with this assignment. 

Suppose the dataset contains grades out of 10 for each assignment and out of 100 for the exam. The grading policy for the course is:

    10% for each assignment and 70% for the final exam.

 

A) Read the data into a dataframe. Is this dataframe tidy? Justify your answer. If "No", then list all ways in which it is not tidy. 

B) Add a new column 'Final', calculated according to the above formula. Put 'INC' (for incomplete) for the student who didn't write the exam. Save the data to a file 'grades_calculated.csv', *without* an initial column containing row numbers (i.e. indexes).

*Hint: run 'df.to_csv?'*

C) Make a graphic showing all individual students' grades for all assessments. 



In [114]:
grades = pd.read_csv("grades.csv")
grades

Name  Asst1     Asst2  Asst3       Exam
0    Adam      9  8.000000    8.0  60.000000
1    Bing      7  8.000000    7.0  75.000000
2   Carla      6       NaN    5.0        NaN
3  Darius      6  7.000000   10.0  80.000000
4    mean      7  7.666667    7.5  71.666667

# A)

The data is not tidy because:
- there are NaN values in the table
- Unless 'mean' is the name of a student, the row at index 4 is innapropriate (A row should only contain observations of variable, which is not the case because 'mean' is not an observation of the variable 'Name')
- This is a wide form representation of the data (The table could just consist of 3 columns : Name, Assignment and Exam)(If we even consider exam and assignments as items of the same category, then the table can even consist of two columns: Name and Grade)




In [205]:
#Here is a long version of the table
long_grades = grades.copy()

pd.melt(long_grades, id_vars=['Name'])

Name variable      value
0     Adam    Asst1   9.000000
1     Bing    Asst1   7.000000
2    Carla    Asst1   6.000000
3   Darius    Asst1   6.000000
4     mean    Asst1   7.000000
5     Adam    Asst2   8.000000
6     Bing    Asst2   8.000000
7    Carla    Asst2        NaN
8   Darius    Asst2   7.000000
9     mean    Asst2   7.666667
10    Adam    Asst3   8.000000
11    Bing    Asst3   7.000000
12   Carla    Asst3   5.000000
13  Darius    Asst3  10.000000
14    mean    Asst3   7.500000
15    Adam     Exam  60.000000
16    Bing     Exam  75.000000
17   Carla     Exam        NaN
18  Darius     Exam  80.000000
19    mean     Exam  71.666667

# B)

In [118]:
asst1 =grades['Asst1']
asst2 =grades['Asst2']
asst3 =grades['Asst3']
exam =grades['Exam']

final = asst1 + asst2 + asst3 + (70*exam/100)

final.loc[final.isnull()] = 'INC'
final

grades_calculated = grades.copy()
grades_calculated['Final'] = final
grades_calculated

Name  Asst1     Asst2  Asst3       Exam      Final
0    Adam      9  8.000000    8.0  60.000000       67.0
1    Bing      7  8.000000    7.0  75.000000       74.5
2   Carla      6       NaN    5.0        NaN        INC
3  Darius      6  7.000000   10.0  80.000000       79.0
4    mean      7  7.666667    7.5  71.666667  72.333333

In [128]:
grades_calculated.to_csv("grades_calculated.csv", index = False)

# C

In [202]:
#tidying the data

to_chart = grades_calculated.copy()
to_chart = to_chart.drop(index = 4)
to_chart.loc[to_chart['Asst2'].isnull(),'Asst2'] = 0
to_chart.loc[to_chart['Exam'].isnull(),'Exam'] = 0
to_chart.loc[to_chart['Final']== 'INC','Final'] = 0

#Clean dataframe to work with
to_chart


Name  Asst1  Asst2  Asst3  Exam Final
0    Adam      9    8.0    8.0  60.0  67.0
1    Bing      7    8.0    7.0  75.0  74.5
2   Carla      6    0.0    5.0   0.0     0
3  Darius      6    7.0   10.0  80.0  79.0

In [203]:

a1_chart = alt.Chart(to_chart).mark_bar().encode(

    x = alt.X('Name', title =""),
    y = alt.Y('Asst1',scale=alt.Scale(domain=[0,10])),
    color = alt.Color('Name', scale=alt.Scale(scheme='pastel1'))
).properties(title='Assignment 1 grades')


a2_chart = alt.Chart(to_chart).mark_bar().encode(

     x = alt.X('Name', title =""),
    y = alt.Y('Asst2',scale=alt.Scale(domain=[0,10])),
    color = 'Name'
).properties(title='Assignment 2 grades')


a3_chart = alt.Chart(to_chart).mark_bar().encode(

     x = alt.X('Name', title =""),
    y = alt.Y('Asst3',scale=alt.Scale(domain=[0,10])),
   color = 'Name'
).properties(title='Assignment 3 grades')

exam_chart = alt.Chart(to_chart).mark_bar().encode(

    x = alt.X('Name', title =""),
    y = alt.Y('Exam',scale=alt.Scale(domain=[0,100])),
    color = 'Name'
).properties(title='Exam grades')

#TODO: Stacked bar chart of the student grades


a1_chart | a2_chart | a3_chart |exam_chart 

alt.HConcatChart(...)

# Question 3 (4 points)

This question uses a larger subset of the Billboard Top 100 sets, in the dataset `billboardhw.csv` provided with this assignment. It is similar to the smaller subset that appeared in Question 6 in the Week 3 Lab.


A) Select all songs that first appeared on the charts in January 2000. (The date of first appearance is 'date_entered'.) 

*Hint: use pd.to_datetime()*

B) Partially tidy this chart, by putting all rankings (for all weeks) in one column named 'ranking'.

*Hint: use pd.melt() rather than wide_to_long().*

C) Produce a single graphic showing, for each song in the chart, a line graph of ranking versus time (week).